src/inmanta/deploy/scheduler.py

```diff
@@ -848,46 +1025,54 @@ async def get_resource_intent(self, resource: ResourceIdStr) -> Optional[Resourc
                 return None
             return ResourceIntent(model_version=self._state.version, details=resource_details, dependencies=None)
 
-    async def deploy_start(self, resource: ResourceIdStr) -> Optional[ResourceIntent]:
+    async def deploy_start(self, action_id: uuid.UUID, resource: ResourceIdStr) -> Optional[ResourceIntent]:
         async with self._scheduler_lock:
             # fetch resource details under lock
             resource_details = self._get_resource_intent(resource)
             if resource_details is None or self._state.resource_state[resource].blocked is BlockedStatus.YES:
+                # We are trying to deploy a stale resource.
                 return None
             dependencies = await self._get_last_non_deploying_state_for_dependencies(resource=resource)
             self._deploying_latest.add(resource)
-            return ResourceIntent(model_version=self._state.version, details=resource_details, dependencies=dependencies)
+            resource_intent = ResourceIntent(
+                model_version=self._state.version, details=resource_details, dependencies=dependencies
+            )
+            # Update the state in the database.
+            await self.send_in_progress(action_id, Id.parse_id(ResourceVersionIdStr(f"{resource},v={self._state.version}")))
+            return resource_intent
 
-    async def deploy_done(
-        self,
-        resource: ResourceIdStr,
-        *,
-        attribute_hash: str,
-        resource_state: const.HandlerResourceState,
+    async def deploy_done(self, attribute_hash: str, result: DeployResult) -> None:
+        deployment_result: DeploymentResult = DeploymentResult.from_handler_resource_state(result.resource_state)
+        try:
+            await self.update_scheduler_state_for_finished_deploy(attribute_hash, result, deployment_result)
+        except StaleResource:
+            # The resource is no longer managed, no need to update the database state.
+            pass
+        else:
+            # Write deployment result to the database.
+            await self.send_deploy_done(attribute_hash, result, deployment_result)
+
+    async def update_scheduler_state_for_finished_deploy(
+        self, attribute_hash: str, result: DeployResult, deployment_result: DeploymentResult
```

Comment:
Any reason you added the `DeploymentResult` back? Merge conflict gone wrong? It was quite intentionally removed. Let me know if you'd like me to look into restoring this method to its intended state.

Reply:
Ok, I see why you did it. The reason it was removed was because it coupled the deployment result and the status (only certain combinations are actually valid). In this implementation the issue exists between result and deployment result. Since it is now an internal method it's not as bad, but could you add to the docstring that the method expects `result` and `deployment_result` to be consistent?

Or well, the conversion is very cheap, so perhaps we can simply do it both withing this method and within the `send_deploy_done`? I think I'd prefer that, but if you don't like it let's stick with documenting it.